src/linear_algebra/matrix/nonsingular_inverse.lean

```diff
@@ -473,15 +540,141 @@ mul_nonsing_inv A (is_unit_det_of_invertible A)
 @[simp] lemma inv_mul_of_invertible [invertible A] : A⁻¹ ⬝ A = 1 :=
 nonsing_inv_mul A (is_unit_det_of_invertible A)
 
+@[simp] lemma inv_zero : (0 : matrix n n α)⁻¹ = 0 :=
+begin
+  casesI (subsingleton_or_nontrivial α) with ht ht,
+  { simp },
+  cases (fintype.card n).zero_le.eq_or_lt with hc hc,
+  { rw [eq_comm, fintype.card_eq_zero_iff] at hc,
+    haveI := hc,
+    ext i,
+    exact (is_empty.false i).elim },
+  { have hn : nonempty n := fintype.card_pos_iff.mp hc,
+    refine nonsing_inv_apply_not_is_unit _ _,
+    simp [hn] },
+end
+
+@[simp] lemma inv_one : (1 : matrix n n α)⁻¹ = 1 :=
+inv_eq_left_inv (by simp)
+
+lemma inv_smul (k : α) [invertible k] (h : is_unit A.det) : (k • A)⁻¹ = ⅟k • A⁻¹ :=
+inv_eq_left_inv (by simp [h, smul_smul])
+
+lemma inv_smul' (k : units α) (h : is_unit A.det) : (k • A)⁻¹ = k⁻¹ • A⁻¹ :=
+inv_eq_left_inv (by simp [h, smul_smul])
+
+lemma _root_.is_unit.coe_inv_mul {α : Type*} [monoid α] {a : α} (h : is_unit a) :
+  ↑(h.unit)⁻¹ * a = 1 :=
+begin
+  convert units.mul_inv _,
+  simp [h.unit_spec]
+end
+
+lemma _root_.is_unit.mul_coe_inv {α : Type*} [monoid α] {a : α} (h : is_unit a) :
+  a * ↑(h.unit)⁻¹ = 1 :=
+begin
+  convert units.mul_inv _,
+  simp [h.unit_spec]
+end
+
+lemma _root_.is_unit.inv_smul {α : Type*} [monoid α] {a : α} (h : is_unit a) :
+  (h.unit)⁻¹ • a = 1 :=
+h.coe_inv_mul
+
+lemma inv_adjugate (A : matrix n n α) (h : is_unit A.det) :
+  (adjugate A)⁻¹ = h.unit⁻¹ • A :=
+begin
+  refine inv_eq_left_inv _,
+  rw [smul_mul, mul_adjugate, units.smul_def, smul_smul, h.coe_inv_mul, one_smul]
+end
+
+@[simp] lemma inv_inv_inv (A : matrix n n α) : A⁻¹⁻¹⁻¹ = A⁻¹ :=
+begin
+  by_cases h : is_unit A.det,
+  { rw [nonsing_inv_nonsing_inv _ h] },
+  { simp [nonsing_inv_apply_not_is_unit _ h] }
+end
+
+lemma mul_inv_rev (A B : matrix n n α) : (A ⬝ B)⁻¹ = B⁻¹ ⬝ A⁻¹ :=
+begin
+  by_cases h : is_unit (A ⬝ B).det,
+  { refine inv_eq_left_inv _,
+    rw det_mul at h,
+    rw [←matrix.mul_assoc, matrix.mul_assoc _ _ A,
+        nonsing_inv_mul _ (is_unit_of_mul_is_unit_left h),
+        matrix.mul_one, nonsing_inv_mul _ (is_unit_of_mul_is_unit_right h)] },
+  { rw nonsing_inv_apply_not_is_unit _ h,
+    rw det_mul at h,
+    have : ¬ is_unit A.det ∨ ¬ is_unit B.det,
+    { contrapose! h,
+      exact h.left.mul h.right },
+    cases this with h' h';
+    simp [nonsing_inv_apply_not_is_unit _ h'] }
+end
+
+lemma ring_hom.map_adjugate {R S : Type*} [comm_ring R] [comm_ring S] (f : R →+* S)
+  (M : matrix n n R) : f.map_matrix M.adjugate = matrix.adjugate (f.map_matrix M) :=
+begin
+  ext i k,
+  have : (λ (j : n), ite (i = j) (1 : S) 0) = f ∘ (λ (j : n), ite (i = j) 1 0),
+  { ext,
+    simp [apply_ite f] },
+  rw [adjugate_apply, ring_hom.map_matrix_apply, map_apply, ring_hom.map_matrix_apply,
+      this, ←map_update_row, ←ring_hom.map_matrix_apply, ←ring_hom.map_det, ←adjugate_apply]
+end
+
+section cancel
+
+-- TOOD: where should this go? general statement about pi.smul?
```

Comment:
Probably `pi.is_smul_regular` should be a theorem high up the import hierarchy (if `algebra/regular/smul.lean` imports `pi.smul` then that would be the obvious place) defined for `pi.smul`, and then we'd have a specialization `matrix.is_smul_regular` which lives in `data/matrix/basic.lean`.